src/rdma.c

```diff
@@ -0,0 +1,985 @@
+
+    /* it's time to tell redis we have already connected */
+    c->flags |= VALKEY_CONNECTED;
+    c->funcs = &valkeyContextRdmaFuncs;
+    c->fd = ctx->comp_channel->fd;
+
+    return connRdmaRegisterRx(ctx, cm_id);
+}
+
+static int ValkeyRdmaCM(valkeyContext *c, int timeout) {
+    RdmaContext *ctx = (RdmaContext *)c->privctx;
+    struct rdma_cm_event *event;
+    char errorstr[128];
+    int ret = VALKEY_ERR;
+
+    while (rdma_get_cm_event(ctx->cm_channel, &event) == 0) {
+        switch (event->event) {
+        case RDMA_CM_EVENT_ADDR_RESOLVED:
+            if (timeout < 0 || timeout > 100)
+                timeout = 100; /* at most 100ms to resolve route */
+            ret = rdma_resolve_route(event->id, timeout);
+            if (ret) {
+                valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: route resolve failed on");
+            }
+            break;
+        case RDMA_CM_EVENT_ROUTE_RESOLVED:
+            ret = ValkeyRdmaConnect(c, event->id);
+            break;
+        case RDMA_CM_EVENT_ESTABLISHED:
+            ret = ValkeyRdmaEstablished(c, event->id);
+            break;
+        case RDMA_CM_EVENT_TIMEWAIT_EXIT:
+            ret = VALKEY_ERR;
+            valkeySetError(c, VALKEY_ERR_TIMEOUT, "RDMA: connect timeout");
+            break;
+        case RDMA_CM_EVENT_ADDR_ERROR:
+        case RDMA_CM_EVENT_ROUTE_ERROR:
+        case RDMA_CM_EVENT_CONNECT_ERROR:
+        case RDMA_CM_EVENT_UNREACHABLE:
+        case RDMA_CM_EVENT_REJECTED:
+        case RDMA_CM_EVENT_DISCONNECTED:
+            c->flags &= ~VALKEY_CONNECTED;
+            break;
+        case RDMA_CM_EVENT_ADDR_CHANGE:
+        default:
+            snprintf(errorstr, sizeof(errorstr), "RDMA: connect failed - %s", rdma_event_str(event->event));
+            valkeySetError(c, VALKEY_ERR_OTHER, errorstr);
+            ret = VALKEY_ERR;
+            break;
+        }
+
+        rdma_ack_cm_event(event);
+    }
+
+    return ret;
+}
+
+static int ValkeyRdmaWaitConn(valkeyContext *c, long timeout) {
+    int timed;
+    struct pollfd pfd;
+    long now = vk_msec_now();
+    long start = now;
+    RdmaContext *ctx = (RdmaContext *)c->privctx;
+
+    while (now - start < timeout) {
+        timed = (int)(timeout - (now - start));
+
+        pfd.fd = ctx->cm_channel->fd;
+        pfd.events = POLLIN;
+        pfd.revents = 0;
+        if (poll(&pfd, 1, timed) < 0) {
+            return VALKEY_ERR;
+        }
+
+        if (ValkeyRdmaCM(c, timed) == VALKEY_ERR) {
+            return VALKEY_ERR;
+        }
+
+        if (c->flags & VALKEY_CONNECTED) {
+            return VALKEY_OK;
+        }
+
+        now = vk_msec_now();
+    }
+
+    return VALKEY_ERR;
+}
+
+static int valkeyContextConnectRdma(valkeyContext *c, const valkeyOptions *options) {
+    const struct timeval *timeout = options->connect_timeout;
+    const char *addr = options->endpoint.tcp.ip;
+    int port = options->endpoint.tcp.port;
+    int ret;
+    char _port[6];  /* strlen("65535"); */
```

Comment:
You're allowed `_port` in local scope but personal I'd prefer a non-underscore name

Reply:
`int port = options->endpoint.tcp.port;` has been already declared as `int` type.